src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -925,301 +1296,410 @@ internal TransitionRegex<TSet> CreateDerivative()
+                            // The loop derivative peels out one iteration and concatenates the body's derivative with the decremented loop,
+                            // so d(R{m,n}) = d(R)R{max(0,m-1),n-1}. Note that n is guaranteed to be greater than zero, since otherwise the
+                            // loop would have been simplified to nothing, and int.MaxValue is treated as infinity.
+                            int newupper = _upper == int.MaxValue ? int.MaxValue : _upper - 1;
+                            int newlower = _lower == 0 ? 0 : _lower - 1;
+                            // the continued loop becomes epsilon when newlower == newupper == 0
+                            // in which case the returned concatenation will be just bodyDerivative
+                            derivative = _builder.CreateConcat(bodyDerivative, _builder.CreateLoop(_left, IsLazy, newlower, newupper));
                         }
+                        break;
                     }
+
+                case SymbolicRegexNodeKind.OrderedOr:
+                    {
+                        Debug.Assert(_left is not null && _right is not null);
+                        derivative = OrderedOr(_builder, _left.CreateDerivative(elem, context), _right.CreateDerivative(elem, context));
+                        break;
+                    }
+
+                case SymbolicRegexNodeKind.Effect:
+                    //Effect nodes do not have derivatives (effects have been stripped): It is an error to reach an Effect node here
+                    Debug.Fail($"{nameof(CreateDerivative)}:{_kind}");
+                    break;
+
+                default:
+                    // The derivative of any other case is nothing
+                    // e.g. taking the derivative of () (epsilon) is [] (nothing)
+                    derivative = _builder._nothing;
                     break;
+            }
+
+            _builder._derivativeCache[key] = derivative;
+            return derivative;
+        }
+
+        /// <summary>
+        /// Remove any Effect nodes from the tree, keeping just the actual pattern.
+        /// So Effect(R,E) would be simplified to just R.
+        /// </summary>
+        /// <returns>the node with all Effect nodes stripped away</returns>
+        internal SymbolicRegexNode<TSet> StripEffects()
+        {
+            if (!StackHelper.TryEnsureSufficientExecutionStack())
+            {
+                return StackHelper.CallOnEmptyStack(StripEffects);
+            }
+
+            // If the node doesn't contain any Effect nodes under it we are done
+            if (!_info.ContainsEffect)
+                return this;
+
+            // Recurse over the structure of the node to strip effects
+            switch (_kind)
+            {
+                case SymbolicRegexNodeKind.Effect:
+                    Debug.Assert(_left is not null);
+                    // This is the place where the effect (the right child) is getting ignored
```

Comment:
I assume _right _is_ null here?

Reply:
Ah no, _right is guaranteed to be not null here (all Effect nodes do have a right child), but it's just not needed. Would asserting _right is not null be clearer?

Reply:
I added the assertion.